source/common/network/listener_impl.cc

```diff
@@ -58,39 +53,50 @@ ListenerImpl::ListenerImpl(Event::DispatcherImpl& dispatcher, Socket& socket, Li
   }
 
   if (bind_to_port) {
-    listener_.reset(
-        evconnlistener_new(&dispatcher.base(), listenCallback, this, 0, -1, socket.fd()));
+    const int ret = ::listen(socket.fd(), 128);
 
-    if (!listener_) {
+    if (ret == -1) {
       throw CreateListenerException(
           fmt::format("cannot listen on socket: {}", socket.localAddress()->asString()));
     }
 
+    file_event_ = dispatcher.createFileEvent(
+        socket.fd(),
+        [this, socket_fd = socket.fd()](uint32_t) {
+          while (true) {
+            sockaddr_storage remote_addr;
+            socklen_t remote_addr_len = sizeof(remote_addr);
+            const int ret =
+                ::accept(socket_fd, reinterpret_cast<sockaddr*>(&remote_addr), &remote_addr_len);
+            if (ret < 0 || ::fcntl(ret, F_SETFL, O_NONBLOCK) == -1) {
+              if (errno == EINTR || errno == EAGAIN || errno == EWOULDBLOCK ||
+                  errno == ECONNABORTED) {
+                return;
+              }
+              PANIC(fmt::format("listener accept failure: {}", strerror(errno)));
+            }
+            listenCallback(ret, reinterpret_cast<sockaddr&>(remote_addr), remote_addr_len);
+          }
+        },
+        Event::FileTriggerType::Level, Event::FileReadyType::Read);
```

Comment:
Is it possible to use edge triggered for accepts? Would that mean we don't have to call `fcntl` on every iteration above. Larger, question (and sorry I need to refresh myself w/ the man pages) where do we set non-blocking in the first iteration?

Reply:
Just following what was done for evconnlistener with edge vs. level. I think you want edge, to allow for enable/disable behavior without losing events.

The fcntl in this loop is for the accepted connection FD, not for the socket FD. The socket FD was  already set non-blocking at https://github.com/envoyproxy/envoy/blob/master/source/common/network/address_impl.cc#L167.

Reply:
To avoid the `fcntl` in the loop, you could use `accept4()`, but I think that is linux-only, so then you'd need some `#ifdefs`. It may be worthwhile in a later change if this is shown to have a meaningful performance impact.

Reply:
ah OK, can we potentially move the `fcntl` to its own call and use RELEASE_ASSERT? I think that should never fail?

Reply:
I agree, this would be clearer with separate error handling for the `fcntl`

Reply:
Yes, I deliberately skipped `accept4` due to the `#ifdef` situation, will leave a TODO.